lib/zephyr/src/zephyr/expr.py

```diff
@@ -0,0 +1,350 @@
+    op: Literal["eq", "ne", "lt", "le", "gt", "ge"]
+
+    def evaluate(self, record: dict) -> bool:
+        left_val = self.left.evaluate(record)
+        right_val = self.right.evaluate(record)
+        return _COMPARE_OPS[self.op](left_val, right_val)
+
+    def __repr__(self) -> str:
+        return f"({self.left} {_COMPARE_SYMBOLS[self.op]} {self.right})"
+
+
+@dataclass(eq=False)
+class LogicalExpr(Expr):
+    """Logical AND/OR between expressions."""
+
+    left: Expr
+    right: Expr
+    op: Literal["and", "or"]
+
+    def evaluate(self, record: dict) -> bool:
+        left_val = self.left.evaluate(record)
+        if self.op == "and":
+            return bool(left_val and self.right.evaluate(record))
+        return bool(left_val or self.right.evaluate(record))
+
+    def __repr__(self) -> str:
+        op_symbol = "&" if self.op == "and" else "|"
+        return f"({self.left} {op_symbol} {self.right})"
+
+
+@dataclass(eq=False)
+class NotExpr(Expr):
+    """Logical NOT."""
+
+    child: Expr
+
+    def evaluate(self, record: dict) -> bool:
+        return not self.child.evaluate(record)
+
+    def __repr__(self) -> str:
+        return f"~{self.child}"
+
+
+_ARITHMETIC_OPS: dict[str, Callable[[Any, Any], Any]] = {
+    "add": operator.add,
+    "sub": operator.sub,
+    "mul": operator.mul,
+    "truediv": operator.truediv,
+}
+
+_ARITHMETIC_SYMBOLS: dict[str, str] = {
+    "add": "+",
+    "sub": "-",
+    "mul": "*",
+    "truediv": "/",
+}
+
+
+@dataclass(eq=False)
+class ArithmeticExpr(Expr):
+    """Arithmetic operation between expressions."""
+
+    left: Expr
+    right: Expr
+    op: Literal["add", "sub", "mul", "truediv"]
+
+    def evaluate(self, record: dict) -> Any:
+        return _ARITHMETIC_OPS[self.op](self.left.evaluate(record), self.right.evaluate(record))
+
+    def __repr__(self) -> str:
+        return f"({self.left} {_ARITHMETIC_SYMBOLS[self.op]} {self.right})"
+
+
+@dataclass(eq=False)
+class FieldAccessExpr(Expr):
+    """Access a field of a nested struct."""
+
+    parent: Expr
+    field: str
+
+    def evaluate(self, record: dict) -> Any:
+        parent_val = self.parent.evaluate(record)
+        if isinstance(parent_val, dict):
+            return parent_val.get(self.field)
+        return None
+
+    def __repr__(self) -> str:
+        return f"{self.parent}[{self.field!r}]"
+
+
+# Convenience functions
+def col(name: str) -> ColumnExpr:
+    """Create a column reference expression.
+
+    Example:
+        >>> col("score") > 0.5
+        (col('score') > lit(0.5))
+    """
+    return ColumnExpr(name)
+
+
+def lit(value: Any) -> LiteralExpr:
+    """Create a literal value expression.
+
+    Example:
+        >>> col("a") + lit(10)
+        (col('a') + lit(10))
+    """
+    return LiteralExpr(value)
+
+
+def expr_to_callable(expr: Expr) -> Callable[[dict], Any]:
+    """Convert an expression to a callable predicate."""
+    return expr.evaluate
+
+
+def extract_columns(expr: Expr) -> set[str]:
```

Comment:
Q: I wonder if it would be cleaner to add `_get_inner_expressions` (or some other name) method to `Expr`, that way it's less likely someone would add an expression without updating `extract_columns`. It would also simplify the `extract_columns`.

Related, we could consider raising if there's expression not handled in the `elif`?

Also this seems like a good use-case for the `match`?

Reply:
this actually was stale code so I removed it entirely, but good idea :)